File: kotlin-code/src/main/kotlin/com/springernature/slides/Slide12.kt
package com.springernature.slides

open class Ship(
    var health: Int = 100
)

interface Gun {
    val strength: Int

    fun fireAt(ship: Ship) {
        ship.health =- (1 * strength)
    }
}

class LaserGun(
    override val strength: Int = 5
): Gun

class Commander(gun: Gun): Ship(), Gun by gun

fun main() {
    val commander = Commander(LaserGun())
    commander.fireAt(Commander(LaserGun()))
    commander.strength
    commander is Gun
}

File: scala-code/src/main/scala/com/springernature/slides/Slide12.scala
package com.springernature.slides

object Slide12 {
  abstract class Ship(var health: Int = 100)

  trait Gun {
    val strength: Int = 1

    def fireAt(ship: Ship) = {
      ship.health -= (1 * strength)
    }
  }

  trait LaserGun extends Gun {
    override val strength: Int = 5
  }

  class Commander extends Ship() with Gun

  def main(args: Array[String]): Unit = {
    val commander = new Commander()
    val laserGunShip = new Ship() with LaserGun
    commander.fireAt(laserGunShip)
    println(commander.strength)
    println(commander.isInstanceOf[Gun])
  }
}
